### DISCUSSION TITLE
还是头疼的换算问题，让我无法自拔

### DISCUSSION THREAD

**Discussion by henry652341974** [smartcontractkit/Web3_tutorial_Chinese#10] [Q&A] [answered]
`
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.20;
import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/shared/interfaces/AggregatorV3Interface.sol";

contract TestDataFee {
        AggregatorV3Interface internal dataFeed =  
        AggregatorV3Interface(0x694AA1769357215DE4FAC081bf1f309aDC325306); 
        uint256 amount;
        // 1. -> 输入 2Wei  msg.value=uint256: 2                    Balance: 0.000000000000000002 ETH
        //            getUSD()=uint256: 5622
        function fund() external payable returns (uint256){
            amount = msg.value;
            return msg.value;
        } 
     
        function getValue() public view returns(uint256){
            return amount;
        }

        function getChainlinkDataFeedLatestAnswer() public view returns (int) {
        (
            /* uint80 roundID */,
            int answer,
            /*uint startedAt*/,
            /*uint timeStamp*/,
            /*uint80 answeredInRound*/
        ) = dataFeed.latestRoundData();
        return answer;
    }   

    function getUSD() public view returns (uint256) {
        return convertEhtToUsd(amount);
    }

     function convertEhtToUsd (uint256 ethAmount) internal view returns (uint256) {
        uint256 ethPrice = uint256(getChainlinkDataFeedLatestAnswer());
        return ethAmount * ethPrice / (10**8);//10**8是ETH对USD的精度.
    }
}
`

---
就这段最简单的代码:
首先我输入 2Wei 
点击getValue()  得到 msg.value=uint256: 2    说明单位就是wei 显示 Balance: 0.000000000000000002 ETH

到此没有问题

当我点击getUSD() uint256: 5622
我想知道这个5622的单位是什么？是USD吗？显然不可能2Wei=5622USD
所以这个5622到底是什么东西？

@QingyangKong 求解~！可能是我太笨了就是转不过弯来

**Comment by henry652341974** [comment]
![Wei转USD](https://github.com/user-attachments/assets/82c953da-14a3-4622-b78e-cef371511c76)

我感觉老师的代码还是有问题

**Comment by henry652341974** [comment]
第三课中的第一个梗：
![image](https://github.com/user-attachments/assets/8239341c-dcbb-4b53-a57f-ee71aff8f373)

di'sa第三课中的第二个梗：
![image](https://github.com/user-attachments/assets/bf078eff-cfac-4e4e-a219-2f3e042a894c)

**Comment by henry652341974** [comment, marked as answer]
我搞 明白了，你把下面该除10的18次方，扔给了MINIMUN_VALUE让他乘10的18次方，所以最终结果还是对的。但老师，你这样讲学生会很懵逼的

**Comment by Charlie-pang-sys** [comment]
我最近也在看17小时这个视频，这块的单位换算让我百思不得其解，反复看了好几遍视频，还担心我自己写错了，把源码翻出来也没看明白，看来有人早都觉得这块不对了，回家我再按照自己的猜想打日志好好验证一下

**Comment by henry652341974** [comment]
我已经看懂了，她是把换算部分放到另外一个逻辑里面了。我们正常的逻辑是先换算。她实际是做了换算，只是做的时机不是在公式里面。确实也不能放在公式里面会出现精度问题。现在这个web3学了还好找工作吗？最近又开始纠结了，加个微信把兄弟，以后慢慢交流。微信就是xuhengyu_com
&nbsp;


/ds…Ⅶ乄、
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2025年2月8日(星期六) 下午5:27
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [smartcontractkit/Web3_tutorial_Chinese] 还是头疼的换算问题，让我无法自拔 (Discussion #10)














 
我最近也在看17小时这个视频，这块让单位换算这块我百思不得其解，反复看了好几遍视频，还担心我自己写错了，把源码翻出来也没看明白，看来有人早都觉得这块不对了，回家我再按照自己的猜想打日志好好验证一下

**Comment by QingyangKong** [comment]
这个问题可能让很多朋友感到很困惑，如果大家把加密货币的基础单位当成是wei而不是ether就很好理解了。

比如说100usd那里，如果写 100，他的单位是 ether，而不是wei。而在 solidity 编程中，默认的单位是wei而非ether，因为如果你的常量单位是 ether，那么在函数中用这个常量来计算，就都需要专门考虑这个常量的单位是 wei还是ether。

打个比方，你在平时编程是，一个变量的值是1，你在定义它的时候会给它复制1，而不会赋值1000乘以10的负三次方。因为在正常的认知中，最小单位是1。在solidity中，把正常的单位调整成为 wei，就很容易理解了。
